cardano-api/test/Test/Cardano/Api/TxBody.hs

```diff
@@ -0,0 +1,285 @@
+--
+-- Roundtrip data requires normalization, too.
+test_roundtrip_TxBody_make_get :: [TestTree]
+test_roundtrip_TxBody_make_get =
+  [ testProperty (show era) $
+    property $ do
+      content <- forAll $ genTxBodyContent era
+      tripping
+        (normalizeOriginal $ review content)
+        (\_ -> makeTransactionBody content)
+        (<&> \(TxBody content') -> normalizeRoundtrip content')
+  | AnyCardanoEra era <- [minBound..]
+  ]
+
+-- | Check that conversion from 'TxBody' to 'TxBodyContent' and back gives
+-- result equivalent to original.
+--
+-- The original randomly-generated data requires
+-- ViewTx/BuildTx type conversion (see Rebuild section).
+--
+-- No normalization is needed here, because 'TxBody' keeps less information
+-- on type and value level than 'TxBodyContent'.
+-- For instance, no special /None/ values.
+test_roundtrip_TxBody_get_make :: [TestTree]
+test_roundtrip_TxBody_get_make =
+  [ testProperty (show era) $
+    property $ do
+      txbody <- forAll $ genTxBody era
+      tripping
+        txbody
+        (\(TxBody content) -> content)
+        (makeTransactionBody . rebuildBodyContent)
+  | AnyCardanoEra era <- [minBound..]
+  ]
+
+
+-- * Normalization
+--
+-- Strip unnecessary details
+--
+-- In many cases, after roundtrip,
+-- @Just mempty@ may become @Nothing@ or vice versa.
+-- Input data also may be generated as either @Just 0@ or @Nothing@.
+-- Order of some items may also change, they need to be reordered.
+
+-- | Normalizations applied to original data only
+normalizeOriginal :: TxBodyContent ViewTx era -> TxBodyContent ViewTx era
+normalizeOriginal content =
+  content
+    { txAuxScripts    = normalizeAuxScripts    $ txAuxScripts    content
+    , txCertificates  = normalizeCertificates  $ txCertificates  content
+    , txIns           = sortOn fst             $ txIns           content
+    , txInsCollateral = normalizeInsCollateral $ txInsCollateral content
+    , txMetadata      = normalizeMetadata      $ txMetadata      content
+    , txMintValue     = normalizeMintValue     $ txMintValue     content
+    , txWithdrawals   = normalizeWithdrawals   $ txWithdrawals   content
+    }
+
+-- | Normalizations applied to roundtrip result data only
+normalizeRoundtrip :: TxBodyContent ViewTx era -> TxBodyContent ViewTx era
+normalizeRoundtrip content@TxBodyContent{txAuxScripts, txIns, txInsCollateral} =
+  content
+    { txAuxScripts    = normalizeAuxScripts    txAuxScripts
+    , txIns           = sortOn fst             txIns
+    , txInsCollateral = normalizeInsCollateral txInsCollateral
+    }
+
+-- | Original data: Unify empty and None.
+--
+-- Roundtrip data: Sort transaction input ids.
+normalizeInsCollateral :: TxInsCollateral era -> TxInsCollateral era
+normalizeInsCollateral = \case
+  TxInsCollateralNone -> TxInsCollateralNone
+  -- for original
+  TxInsCollateral _ ins | null ins -> TxInsCollateralNone
+  -- for roundtrip
+  TxInsCollateral support ins -> TxInsCollateral support $ sort ins
```

Comment:
We could refactor as follows:
```
  TxInsCollateral support ins
    | null ins -> TxInsCollateralNone
    | otherwise -> TxInsCollateral support $ sort ins
 ```